index.js

```diff
@@ -0,0 +1,179 @@
+const uniqIngredients2 = products.reduce((acc, { ingredients }) => [...new Set([...acc, ...ingredients])],[]);
+
+//third way reduce/filter
+//prettier-ignore
+const uniqIngredients3 = products.reduce((acc, { ingredients }) => {
+  return [...acc, ...ingredients].filter((element, i, self) => self.indexOf(element) === i);
+}, []);
+
+//2. Создать функцию, которая принимает массив продуктов и id, и возвращает продукт с таким же id.
+const filterByIdFunction = (enterArray, id) =>
+  enterArray.filter((element) => element.id === id);
+
+//3. Создать массив с отсортированными продуктами по цене.
+const coastPrice = products
+  .map((product) => product.price)
+  .sort((a, b) => a - b);
+
+//4. Сгруппировать продукты по типам. Создать объект, где ключ это тип, а значение - массив с продуктами.
+const groupProductsByType = products.reduce((acc, product) => {
+  const { type } = product;
+  return { ...acc, [type]: [...(acc[type] || []), product] };
+}, {});
+
+//5. Создать массив с продуктами, которые доступны.
+//prettier-ignore
+const filterByAvailable = products.filter(product=>product.available===true)
+
+//6. Создать функцию, которая принимает массив продуктов и строку = название ингредиента, и возвращает массив с продуктами, где содержится такой ингредиент.
+//prettier-ignore
+const findProductByIngredient = (enterArray, substring) => products.filter((product) => product.ingredients.includes(substring));
+
+//7. Создать функцию, которая принимает массив продуктов и массив ингредиентов, и возвращает массив с продуктами, где содержатся такие ингредиенты.
+const findProductByIngredients = (enterArray, enterIngredients) => {
+  return enterArray.reduce((acc, product) => {
+    const { ingredients } = product;
+    return ingredients.some((el) => enterIngredients.indexOf(el) >= 0)
+      ? [...acc, product]
+      : [...acc];
+  }, []);
+};
+
+//8. Создать функцию, которая принимает массив продуктов и цену, и возвращает массив продуктов, где цена продукта ниже или равна цене из второго аргумента функции.
+const filterProductsByPrice = (enterArray, price) => {
+  return enterArray.filter((product) => product.price <= price);
+};
+
+//9. Создать функцию, которая принимает массив продуктов и массив айдишников, и возвращает строку, где строка включает в себя название продуктов и их цену через запятую, у которых айдишники совпадают.
+const stringFromProducts = (enterArray, enterId) => {
+  return enterArray.reduce((acc, product) => {
+    const { id, name, price, currency } = product;
+    return enterId.includes(id)
+      ? (acc += `${name}:${price} ${currency}, `)
+      : acc;
+  }, "");
+};
+
+//10. Создать функцию, которая принимает массив продуктов и массив айдишников, и возвращает объект, c общими суммами цен продуктов(у которых айдишники совпадают) по каждой валюте.
+const objectFromProductsByCurrency = (enterArray, enterId) => {
+  return enterArray.reduce(
+    (acc, product) => {
+      const { id, price, currency } = product;
```

Comment:
const hasProduct = enterId.includes(id)  and use it in the next conditions
why? because enterId.includes(id) - 2 times